linear_model.py

```diff
@@ -0,0 +1,118 @@
+
+import os
+import sys
+
+import tensorflow as tf
+
+class LinearModel:
+
+    def __init__(self, input_size, output_size, checkpoint=None):
+        self.input_size = input_size
+        self.output_size = output_size
+
+        self.sess = tf.Session()
+        self._build()
+        self.saver = tf.train.Saver()
+
+        if checkpoint is not None:
+            self.restore_model(checkpoint)
+        else:
+            self.sess.run(tf.global_variables_initializer())
+
+        self.iter_number = 0
+
+    ## Public methods
+
+    def predict(self, state):
+        """
+        state: B x I
+        """
+        output = self.sess.run(self.out, feed_dict={self.input_state: state})
+        return output
+
+    def train_reinforce(self, state, pred, action, var, reward):
+        """
+        B = batch size
+        I = input size
+        A = action dimension
+
+        state: B x I    input state being trained
+        pred: B x A     output from model.predict(state)
+        action: B x A   sampled action using mean from pred
+        var: B x A      variance used for sampling action (MUST BE POSTIIVE)
+        reward: B x 1   reward from environment for the performed action
+
+        inputs expected to be numpy arrays
+        """
+
+        # TODO: for actor-critic, pred should be predicted from a separate critic network instead
+
+        target = (action - pred) * reward / (2.0 * var)
```

Comment:
Just recreating the derivation to confirm:

We want the derivative of -(pred - target)^2 to equal (action - pred) * r / var * grad(pred). (NOTE: there's a negative because optimizers are minimizers, which means they SUBTRACT the gradient from the parameter. However, from the lecture notes it looks like the policy gradient is ADDING to the parameter (see MC policy gradient here))

The derivative is -2(pred - target)grad(pred) = (action - pred) * r / var. Therefore target should be

target = [(action - pred) * r / (2var)] + pred

I'm not very confident on this but it seems to confirm what you are saying. I've updated the code.